src/cryptography/x509/ocsp.py

```diff
@@ -184,279 +137,339 @@ def __init__(
         """
 
     @abc.abstractproperty
-    def issuer_key_hash(self):
+    def issuer_key_hash(self) -> bytes:
         """
         The hash of the issuer public key
         """
 
     @abc.abstractproperty
-    def issuer_name_hash(self):
+    def issuer_name_hash(self) -> bytes:
         """
         The hash of the issuer name
         """
 
     @abc.abstractproperty
-    def hash_algorithm(self):
+    def hash_algorithm(self) -> hashes.HashAlgorithm:
         """
         The hash algorithm used in the issuer name and key hashes
         """
 
     @abc.abstractproperty
-    def serial_number(self):
+    def serial_number(self) -> int:
         """
         The serial number of the cert whose status is being checked
         """
 
     @abc.abstractproperty
-    def extensions(self):
+    def extensions(self) -> x509.Extensions:
         """
         The list of response extensions. Not single response extensions.
         """
 
     @abc.abstractproperty
-    def single_extensions(self):
+    def single_extensions(self) -> x509.Extensions:
         """
         The list of single response extensions. Not response extensions.
         """
+
+
+class OCSPRequestBuilder(object):
+    def __init__(self, request=None, extensions=[]):
+        self._request = request
+        self._extensions = extensions
+
+    def add_certificate(
+        self,
+        cert: x509.Certificate,
+        issuer: x509.Certificate,
+        algorithm: hashes.HashAlgorithm,
+    ):
+        if self._request is not None:
+            raise ValueError("Only one certificate can be added to a request")
+
+        _verify_algorithm(algorithm)
+        if not isinstance(cert, x509.Certificate) or not isinstance(
+            issuer, x509.Certificate
+        ):
+            raise TypeError("cert and issuer must be a Certificate")
+
+        return OCSPRequestBuilder((cert, issuer, algorithm), self._extensions)
+
+    def add_extension(self, extval: x509.ExtensionType, critical: bool):
+        if not isinstance(extval, x509.ExtensionType):
+            raise TypeError("extension must be an ExtensionType")
+
+        extension = x509.Extension(extval.oid, critical, extval)
+        _reject_duplicate_extension(extension, self._extensions)
+
+        return OCSPRequestBuilder(
+            self._request, self._extensions + [extension]
+        )
+
+    def build(self) -> OCSPRequest:
+        from cryptography.hazmat.backends.openssl.backend import backend
+
+        if self._request is None:
+            raise ValueError("You must add a certificate before building")
+
+        return backend.create_ocsp_request(self)
+
+
+class OCSPResponseBuilder(object):
+    def __init__(
+        self, response=None, responder_id=None, certs=None, extensions=[]
+    ):
+        self._response = response
+        self._responder_id = responder_id
+        self._certs = certs
+        self._extensions = extensions
+
+    def add_response(
```

Comment:
All these functions should have `-> "OCSPResponseBuilder"` for return type so mypy can do chaining.